### PR TITLE
[2.4] XGBoost User Experience Improvement

### PR DESCRIPTION
**Pull request by nvidianz**
### Description

All new XGBoost improvement made in main branch is cherry-picked to this release.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Review by yanchengnv — DISMISSED** [review]
Please do another round of e2e test.

**Review by yanchengnv — APPROVED** [review]
Approved

File: integration/xgboost/processor/src/dam/dam.cc [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    size += 8*value.size();
+    entries_->push_back(new Entry(kDataTypeBufferArray, reinterpret_cast<const uint8_t *>(&value), size));
+}
+
+
+std::uint8_t * DamEncoder::Finish(size_t &size) {
+    encoded_ = true;
+
+    size = CalculateSize();
+    auto buf = static_cast<uint8_t *>(calloc(size, 1));
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/dam/dam.cc:108]: Make sure this buf is freed eventually. Maybe we should let the Encoder remember all allocated objects, and free them in the Finish method. This way, others won't need to deal with free.
But we must make sure the Finish method is always called!

File: integration/xgboost/processor/src/include/nvflare_processor.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include <string>
+#include <vector>
+#include <map>
+#include "processing/processor.h"
+#include "data_set_ids.h"
+
+class NVFlareProcessor: public processing::Processor {
+ private:
+    bool active_ = false;
+    const std::map<std::string, std::string> *params_;
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/include/nvflare_processor.h:27]: How is *params_ allocated and freed?

File: integration/xgboost/processor/src/include/nvflare_processor.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        this->active_ = active;
+    }
+
+    void Shutdown() override {
+        this->gh_pairs_ = nullptr;
+        this->cuts_.clear();
+        this->slots_.clear();
+    }
+
+    void FreeBuffer(void *buffer) override {
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/include/nvflare_processor.h:45]: What buffer is this intended to free? Please check and add description.

File: integration/xgboost/processor/src/nvflare-plugin/local_mock.cc [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+#include "local_mock.h"
+#include "data_set_ids.h"
+
+void* LocalMockProcessor::ProcessHistograms(std::size_t *size, const std::vector<double>& histograms) {
+    if (debug_) {
+        std::cout << "ProcessHistograms called with " << histograms.size() << " entries" << std::endl;
+    }
+    DamEncoder encoder(kDataSetHistogramResult, true);
+    encoder.AddFloatArray(histograms);
+    return encoder.Finish(*size);
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/local_mock.cc:26]: This func returns a dynamically allocated buffer, who is responsible for freeing it?

File: integration/xgboost/processor/src/nvflare-plugin/local_mock.cc [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    return result;
+}
+
+Buffer LocalMockProcessor::EncryptVector(const std::vector<double>& cleartext) {
+    if (debug_) {
+        std::cout << "Encrypt vector size: " << cleartext.size() << std::endl;
+    }
+
+    size_t size = cleartext.size() * 8;
+    auto buf = malloc(size);
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/local_mock.cc:68]: This is dynamically allocated. Where is it freed?

File: integration/xgboost/processor/src/nvflare-plugin/local_processor.cc [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+void LocalProcessor::Shutdown() {
+    gh_pairs_ = nullptr;
+    FreeEncryptedData(encrypted_gh_);
+    delete histo_;
+    cuts_.clear();
+    slots_.clear();
+}
+
+void LocalProcessor::FreeBuffer(void *buffer) {
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/local_processor.cc:31]: What buffer is intended to be freed here?

File: integration/xgboost/processor/src/nvflare-plugin/local_processor.cc [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+    DamEncoder encoder(kDataSetGHPairs, true, dam_debug_);
+    encoder.AddBuffer(encrypted_data);
+    auto buffer = encoder.Finish(*size);
+    FreeEncryptedData(encrypted_data);
+
+    // Save pairs for future operations. This is only called on active site
+    gh_pairs_ = new std::vector<double>(pairs);
+
+    return buffer;
```
Comment (yanchengnv) [inline integration/xgboost/processor/src/nvflare-plugin/local_processor.cc:61]: This buffer is dynamically allocated, and must be freed after use.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Please see comments for improvemnet.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
See comments

File: integration/xgboost/processor/src/dam/dam.cc [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+void  DamEncoder::AddBuffer(const Buffer &buffer) {
+    if (debug_) {
+        std::cout << "AddBuffer called, size:  " << buffer.buf_size << std::endl;
+    }
+    if (encoded_) {
+        std::cout << "Buffer is already encoded" << std::endl;
+        return;
+    }
+    // print_buffer(buffer, buf_size);
+    entries_->push_back(new Entry(kDataTypeBuffer, reinterpret_cast<const uint8_t *>(buffer.buffer), buffer.buf_size));
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/dam/dam.cc:57]: Changed to object.

File: integration/xgboost/processor/src/dam/dam.cc [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    if (debug_) {
+        std::cout << "AddIntArray called, size:  " << value.size() << std::endl;
+    }
+
+    if (encoded_) {
+        std::cout << "Buffer is already encoded" << std::endl;
+        return;
+    }
+    // print_buffer(buffer, buf_size);
+    entries_->push_back(new Entry(kDataTypeIntArray, reinterpret_cast<const uint8_t *>(value.data()), value.size()));
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/dam/dam.cc:83]: Changed to object.

File: integration/xgboost/processor/src/dam/dam.cc [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+    if (debug_) {
+        std::cout << "AddFloatArray called, size:  " << value.size() << std::endl;
+    }
+
+    if (encoded_) {
+        std::cout << "Buffer is already encoded" << std::endl;
+        return;
+    }
+    // print_buffer(reinterpret_cast<uint8_t *>(value.data()), value.size() * 8);
+    entries_->push_back(new Entry(kDataTypeFloatArray, reinterpret_cast<const uint8_t *>(value.data()), value.size()));
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/dam/dam.cc:70]: Changed to object.

File: integration/xgboost/processor/src/dam/dam.cc [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    if (encoded_) {
+        std::cout << "Buffer is already encoded" << std::endl;
+        return;
+    }
+    size_t size = 0;
+    for (auto &buf: value) {
+        size += buf.buf_size;
+    }
+    size += 8*value.size();
+    entries_->push_back(new Entry(kDataTypeBufferArray, reinterpret_cast<const uint8_t *>(&value), size));
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/dam/dam.cc:100]: Changed to object.

File: integration/xgboost/processor/src/dam/dam.cc [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+            memcpy(pointer, &entry->size, 8);
+            pointer += 8;
+            len = entry->size * entry->ItemSize();
+            if (len) {
+                memcpy(pointer, entry->pointer, len);
+            }
+        }
+        pointer += align(len);
+    }
+
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/dam/dam.cc:151]: Changed to object, no more allocated entries.

File: integration/xgboost/processor/src/include/nvflare_processor.h (hunk 3 above)
Comment (nvidianz) [reply integration/xgboost/processor/src/include/nvflare_processor.h:45]: This is intended to free allocated buffer returned API call.

File: integration/xgboost/processor/src/nvflare-plugin/local_mock.cc (hunk 4 above)
Comment (nvidianz) [reply integration/xgboost/processor/src/nvflare-plugin/local_mock.cc:26]: This is freed in ProcessGHPairs().

File: integration/xgboost/processor/src/nvflare-plugin/local_mock.cc (hunk 5 above)
Comment (nvidianz) [reply integration/xgboost/processor/src/nvflare-plugin/local_mock.cc:68]: This is freed in ProcessGHPairs in local_processor.cc

File: integration/xgboost/processor/src/nvflare-plugin/local_processor.cc (hunk 6 above)
Comment (nvidianz) [reply integration/xgboost/processor/src/nvflare-plugin/local_processor.cc:31]: This is called by XGBoost code. All processor interface returned memory should be freed with this call.

File: integration/xgboost/processor/src/nvflare-plugin/local_processor.cc (hunk 7 above)
Comment (nvidianz) [reply integration/xgboost/processor/src/nvflare-plugin/local_processor.cc:61]: Changed to object, not a pointer.

File: integration/xgboost/processor/src/nvflare-plugin/nvflare_processor.cc [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+    // For each node, get the rows
+    auto row_ids = vector<vector<int64_t>>(nodes.size());
+    int count = 0;
+    for (const auto &kv : nodes) {
+        auto& rows = row_ids[count];
+        for (auto row : kv.second) {
+            rows.push_back(row);
+        }
+        encoder.AddIntArray(rows);
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply integration/xgboost/processor/src/nvflare-plugin/nvflare_processor.cc:108]: Good catch. This was a problem. But now the rows is a reference to row_ids which is allocated out of this block.

**Review by yanchengnv — APPROVED** [review]
LGTM

**Pull request merged**